glommio/src/executor/mod.rs

```diff
@@ -572,75 +579,145 @@ impl Default for LocalExecutorBuilder {
+/// `MaxSpreadOnCpus`.
+///
+/// ```
+/// use glommio::{CpuSet, LocalExecutorPoolBuilder, Placement};
+///
+/// let cpus = CpuSet::online()
+///     .expect("Err: please file an issue with glommio")
+///     .filter(|l| l.numa_node == 0)
+///     .filter(|l| l.package % 2 == 0);
+///
+/// let handles = LocalExecutorPoolBuilder::new(4)
+///     .placement(Placement::MaxSpreadOnCpus(cpus))
+///     .on_all_shards(|| async move {
+///         // ... important stuff ...
+///     })
+///     .unwrap();
+///
+/// handles.join_all();
+/// ```
 #[derive(Debug)]
 pub enum Placement {
     /// For the `Unbound` variant, the [`LocalExecutor`]s created by a
-    /// [`LocalExecutorPoolBuilder`] are not bound to any CPU.
+    /// [`LocalExecutorPoolBuilder`] are not bound to any CPU.  This is the
+    /// default placement.
     Unbound,
-    /* MaxSpread,
-     * MaxPack,
+    /// The `SharedOnCpus` variant binds each [`LocalExecutor`] to the set of
+    /// CPUs specified by [`CpuSet`].  With an unfiltered CPU
+    /// set returned by [`CpuSet::online`], this is similar to using `Unbound`
+    /// with the distinction that bringing additional CPUs online will not
+    /// allow the executors to run on the newly available CPUs.  The
+    /// `SharedOnCpus` variant allows the number of shards specified in
+    /// [`LocalExecutorPoolBuilder::new`] to be greater than the number of CPUs
+    /// as long as at least one CPU is included in `CpuSet`.
+    ///
+    /// #### Errors
+    ///
+    /// If the provided [`CpuSet`] contains no CPUs, a call to
+    /// [`LocalExecutorPoolBuilder::on_all_shards`] will return `Result::
+    /// Err`.
+    SharedOnCpus(CpuSet),
+    /// Each [`LocalExecutor`] is pinned to a particular [`CpuLocation`] such
+    /// that the set of all CPUs selected has a high degree of sepration.
+    /// The selection proceeds from all CPUs that are online in a
+    /// non-deterministic manner.  See [`Placement::MaxSpreadOnCpus`] for a
+    /// way to restrict the set of CPUs available.
+    ///
+    /// #### Errors
+    ///
+    /// If the number of shards specified in [`LocalExecutorPoolBuilder::new`]
+    /// is greater than the number of CPUs available, then a call to
+    /// [`LocalExecutorPoolBuilder::on_all_shards`] will return `Result::
+    /// Err`.
+    MaxSpread,
+    /// Like [`Placement::MaxSpread`] but allows manually restricting the CPUs
+    /// on which [`LocalExecutor`]s can be placed via a [`CpuSet`].  If the
+    /// number of shards specified in [`LocalExecutorPoolBuilder::new`]
+    /// is less than the number of CPUs in the provided `CpuSet` then some
+    /// CPUs will not be used.
+    ///
+    /// #### Errors
+    ///
+    /// If the number of shards specified in [`LocalExecutorPoolBuilder::new`]
+    /// is greater than the number of CPUs available, then a call to
+    /// [`LocalExecutorPoolBuilder::on_all_shards`] will return `Result::
+    /// Err`.
+    MaxSpreadOnCpus(CpuSet),
```

Comment:
What is the justification for the `OnCpus` in your opinion?
My first impression for this is that we could just have `MaxSpread` take an `Option`:
if the `Option` is `None`, behave like your current `MaxSpread`, and if the `Option` contains `Some(CpuSet)`, then abide by that

Your placement code would then be:

```
            Placement::MaxSpread(cpus) => {
                let cpus = cpus.unwrap_or(CpuSet::online()?);
                Self::check_nr_cpus(nr_shards, &cpus)?;
                Self::MaxSpread(MaxSpreader::from_cpu_set(cpus))
            }
```

Reply:
Caution -- somewhat lengthy:

I initially set up `Placement` as:

```rust
pub enum Placement {
    Unbound(Option<CpuSet>),
    MaxSpread(Option<CpuSet>),
    MaxPack(Option<CpuSet>),
}
```

But doing the implementation and docs it felt like `Unbound(Option<CpuSet>)` never worked particularly well.  Specifically, the implementation of `Unbound(None)` was unclear:  should this parse `sysfs` and then bind to all the CPUs that were online (implying that the builder can fail if we have some problem parsing `sysfs`), or should we simply ignore `sysfs` altogether.  Ignoring `sysfs` has two benefits:  1) if we fail to parse `sysfs` for some reason, at least `Unbound` would still work and 2) if a user brings additional CPUs online after starting the executors, they would be able to run on the newly online CPUs.  However, documenting this behavior for `Unbound(None)` felt a little clumsy, in particular in explaining that `Unbound(Some(all_online_cpus))` is different from `Unbound(None)`.  This compares to having the same behavior for e.g. `MaxSpread(None)` and `MaxSpread(Some(all_online_cpus))`.

Perhaps we should have:

```rust
pub enum Placement {
    Unbound,
    Shared(Option<CpuSet>),
    MaxSpread(Option<CpuSet>),
    MaxPack(Option<CpuSet>),
}
```

I went with the current implementation because writing `MaxSpread` just felt a little nicer than `MaxSpread(None)` and I didn't mind `MaxSpreadOnCpus(cpu_set)` relative to `MaxSpread(Some(cpu_set))`.  I suppose it makes it slightly less clear that the underlying implementation is the same, i.e. `MaxSpread` and `MaxSpreadOnCpus(all_online_cpus)` are identical.  I didn't think `Shared` a.k.a. `Shared(None)` would get much use when we already have `Unbound`, so I didn't include it as companion to `SharedOnCpus`, but maybe should have just for completeness.

Long story short: I'm happy to go with anything that doesn't involve `Unbound(Option<CpuSet>)`.  Let me know what you prefer.

Reply:
First, for me the behavior that `Unbound(None)` should have is very clear: it means you really don't care, so you don't parse sysfs at all, and never even call any syscall. It is the same as today: it's whatever the OS wants.

I see semantic value in having that as a separate option, though.

I still don't see what you mean by "Shared". Why "Shared"  ? (naming things: the hardest problem in computer science!!)

but regardless of the name, I like what you proposed last with Unbound being separate, and I agree its `None` version makes little sense, so I'd refine this to:

```rust
pub enum Placement {
    Unbound, // don't do anything
    Shared(CpuSet), // restrict to the following cpuset. Because you're always specifying it, there's no point in an option
    MaxSpread(Option<CpuSet>), // None => MaxSpread on all avail CPU. Set: MaxSpread in the given cpuset
    MaxPack(Option<CpuSet>), // ditto
}
```

Reply:
That works for me.  There might be a better name than `Shared` -- maybe it'll come to us.  I went with `Shared` because the same `CpuSet` is shared amongst all the executors and each of the CPUs has a shared responsibility for all executors ... `AllForOneAndOneForAll(CpuSet)` / `UnusProOmnibusOmnesProUno(CpuSet)`?

Reply:
I don't know, after the `ScopedTask`, @bryandmc became my go-to person for naming things lol.
Maybe he has an opinion here.

However I think the name should say something about the CPU placement. The fact that it is shared across executors is not that relevant.

What we are looking for is something that communicates the fact that it is restricted to a subset of all cpus, but free to move within that set.

Maybe `Restricted`, `Subset`, or something like that ?